include/ginkgo/core/base/machine_topology.hpp

```diff
@@ -0,0 +1,450 @@
+     *
+     * @return  the number of PUs.
+     */
+    size_type get_num_pus() const { return std::get<0>(this->pus_).size(); }
+
+    /**
+     * Get the number of core objects stored in this Topology tree.
+     *
+     * @return  the number of cores.
+     */
+    size_type get_num_cores() const { return std::get<0>(this->cores_).size(); }
+
+    /**
+     * Get the number of PCI device objects stored in this Topology tree.
+     *
+     * @return  the number of PCI devices.
+     */
+    size_type get_num_pci_devices() const { return this->pci_devices_.size(); }
+
+    /**
+     * Get the number of NUMA objects stored in this Topology tree.
+     *
+     * @return  the number of NUMA objects.
+     */
+    size_type get_num_numas() const { return this->num_numas_; }
+
+protected:
+    /**
+     * Creates a new MachineTopology object.
+     */
+    static std::shared_ptr<MachineTopology> create()
+    {
+        return std::shared_ptr<MachineTopology>(new MachineTopology());
+    }
+
+    /**
+     * The default constructor that loads the different objects.
+     */
+    MachineTopology();
+
+    /**
+     * @internal
+     *
+     * A helper function that binds the object with an id.
+     */
+    void hwloc_binding_helper(
+        const std::tuple<std::vector<MachineTopology::normal_obj_info>,
+                         MachineTopology::hwloc_manager<hwloc_bitmap_s>> &obj,
+        const int *id, const size_type num_ids,
+        const BitMapType &bitmap_type) const;
+
+    /**
+     * @internal
+     *
+     * A helper function that prints the topology tree of an object to a given
+     * depth. Provided from the hwloc library.
+     */
+    void hwloc_print_children(const hwloc_obj_t obj, const int depth = 1);
+
+    /**
+     * @internal
+     *
+     * Load the objects of a normal HWLOC type (Packages, cores, numa-nodes).
+     *
+     * @note The objects should be sorted by logical index since hwloc uses
+     * logical index with these functions
+     */
+    void load_objects(hwloc_obj_type_t type,
+                      std::tuple<std::vector<normal_obj_info>,
+                                 hwloc_manager<hwloc_bitmap_s>> &objects);
+
+    /**
+     * @internal
+     *
+     * Load the objects of io type (PCI devices and OS devices).
+     *
+     * @note The objects should be sorted by logical index since hwloc uses
+     * logical index with these functions
+     */
+    void load_objects(hwloc_obj_type_t type, std::vector<io_obj_info> &vector);
+
+    /**
+     * @internal
+     *
+     * Initialize the topology object.
+     */
+    hwloc_topology *init_topology();
+
+    /**
+     * @internal
+     *
+     * Initialize the bitmap.
+     */
+    hwloc_bitmap_s *init_bitmap();
+
+    /**
+     *
+     * @internal
+     *
+     * Get local id from the os index
+     */
+    inline int get_obj_local_id_by_os_index(
+        std::vector<normal_obj_info> &objects, size_type os_index) const;
```

Comment:
Rather than by os_index, we maybe want the logical index one instead?

Reply:
I think binding is always done with the OS index and hence I need the `os_index` [HWLOC reference](https://www.open-mpi.org/projects/hwloc/doc/v2.4.0/a00175.php#ga062dbff93baeff3b425a7260c5463646)